### DISCUSSION TITLE
SUPPORT WITH ERROR

### DISCUSSION THREAD

**Discussion by FabioNotaro2001** [NVIDIA/NVFlare#1444] [Q&A]
Hi everybody, when I launch a training (it happens both with hello-pt-tb and with a customized example too) after I submit the job an error occours into the server's terminal basically saying that in file pt_model_persistor.py (that I didn't modified) line 199 it expected a torch.nn.Module but instead he got a class dict.

What can it be? How can I resolve? 
![Schermata del 2023-03-02 16-25-52](https://user-images.githubusercontent.com/93600518/222472577-dc49f999-6d47-4f65-b823-7cc45acc0915.png)

**Comment by chesterxgchen** [comment]
@FabioNotaro2001 

which version of nvflare are you using ? and which branch of the examples are you using ? 
The github default branch is "dev" which is unstable version we are actively changing ( including examples), the stable branch are release branches and main (latest release branch). 
if you did the 
pip install nvflare 

which will install the nvflare==2.2.5 (https://pypi.org/project/nvflare/) 
if you want to use the match examples, 

you should do the followings
git clone <NVFLARE repo>
git switch main 

If you are install the specific version of NVFLARE, you need to switch the match release branch to use the match examples. 

I am not sure this is your issue, but at least this will help rule out some of the basic issues. 

If you did all the right things. This still occurs, I am sure we can help

**Reply by FabioNotaro2001** [reply]
I was actually in the wrong branch, so I ran the git switch main command and also git pull on the server (since on the clients I didn't clone the repository) but when re-running the training appears the exact same error. 

I also tried recreating a virtual environment from scratch but still didn't solve, so I also tried recreating a new virtual environment on a new pc, but whatever I do I get that strange error

**Comment by YuanTingHsieh** [comment]
@FabioNotaro2001 
Did you install the required dependencies specified in the README of those examples?
For hello-pt-tb example in 2.2, you will need to do this for all machines (servers/clients): `pip install torch torchvision tensorboard`

**Reply by FabioNotaro2001** [reply]
Yes, I had already installed those dependencies. To be sure, I re-executed that command but unfortunately the error remains

**Reply by chesterxgchen** [reply]
it could be some initialization issue with your customer class, but we swallow the exception ( this is a bug we are trying to fix), so you see the strange configuration error. 
please double check if the you class arguments match with the configuration specified (args)

**Reply by FabioNotaro2001** [reply]
Could you be more precise please? What do I have to check? In which file?

Effectively, in my custom folder I changed some things. I don't use anymore the SimpleNetwork into the pt_learner.py file but a customized model. In any case I have already updated the file config_fed_server.json by specifying my custom model instead of the default SimpleNetwork parameter. 
Thank you for your patience.

**Reply by chesterxgchen** [reply]
In you customer class, MyExecutor or MyLeaner, you do something like this

  class MyLeaner: 
      def __init__(self, x:str, y:int, z:dict): 
             <do whatever>
     def train() : 
           do something. 

the corresponding config in fed_client_config.json
    you will have components configuration, like
 "components": [
    {
      "id": "my_learner",
      "path": "my.dummy.foo.package.MyLeaner",
      "args": {
        "x": "X", 
        "y: 2, 
       "z": { 
             "a": 1, 
             "b": 2
           }, 
          "t": 123
     
      }
    }
  ]
   
Notice the MyLearner only need 3 arguments: x,y,z, but I put a t = 123 there. which will instantiate class MyLeaner(x,y,z, t), which will fail. 
Unfortunately, the current release, the exception is swallowed, so it bring some configuration as this happened during configuration loading and component construction. We fixed this in the next release (release soon). 
But I am not sure if you issue is this issue. If above still not the solve your issue. If you don't mind, you can share your code with us, we might be able to spot something

**Comment by YuanTingHsieh** [comment]
@FabioNotaro2001 I'll close this discussion, feel free to open a new one if you have any questions!
